File: plugin/openapi/src/main/java/com/gatehill/imposter/plugin/openapi/OpenApiPluginImpl.java
package com.gatehill.imposter.plugin.openapi;

import com.fasterxml.jackson.core.JsonGenerationException;
import com.gatehill.imposter.ImposterConfig;
import com.gatehill.imposter.plugin.ScriptedPlugin;
import com.gatehill.imposter.plugin.config.ConfiguredPlugin;
import com.gatehill.imposter.plugin.openapi.service.OpenApiService;
import com.gatehill.imposter.script.ResponseBehaviour;
import com.gatehill.imposter.util.HttpUtil;
import com.gatehill.imposter.util.MapUtil;
import com.google.common.base.Strings;
import com.google.common.cache.Cache;
import com.google.common.cache.CacheBuilder;
import com.google.common.collect.Lists;
import com.google.common.collect.Maps;
import com.google.common.collect.Sets;
import io.swagger.models.*;
import io.swagger.parser.SwaggerParser;
import io.vertx.core.Handler;
import io.vertx.core.http.HttpMethod;
import io.vertx.core.http.HttpServerResponse;
import io.vertx.ext.web.Router;
import io.vertx.ext.web.RoutingContext;
import io.vertx.ext.web.handler.StaticHandler;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import javax.inject.Inject;
import java.nio.file.Paths;
import java.util.*;
import java.util.concurrent.ExecutionException;
import java.util.function.BiConsumer;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

import static com.gatehill.imposter.util.AsyncUtil.handleAsync;
import static com.gatehill.imposter.util.HttpUtil.CONTENT_TYPE;
import static com.gatehill.imposter.util.HttpUtil.CONTENT_TYPE_JSON;
import static java.util.Optional.empty;
import static java.util.Optional.ofNullable;

/**
 * Plugin for OpenAPI (OAI; formerly known as 'Swagger').
 *
 * @author Pete Cornish {@literal <outofcoffee@gmail.com>}
 */
public class OpenApiPluginImpl extends ConfiguredPlugin<OpenApiPluginConfig> implements ScriptedPlugin<OpenApiPluginConfig> {
    private static final Logger LOGGER = LogManager.getLogger(OpenApiPluginImpl.class);
    private static final Pattern PATH_PARAM_PLACEHOLDER = Pattern.compile("\\{([a-zA-Z]+)\\}");
    private static final String UI_WEB_ROOT = "swagger-ui";
    private static final String ARG_BASEPATH = "openapi.basepath";
    private static final String ARG_SCHEME = "openapi.scheme";
    private static final String ARG_TITLE = "openapi.title";
    static final String SPECIFICATION_PATH = "/_spec";
    static final String COMBINED_SPECIFICATION_PATH = SPECIFICATION_PATH + "/combined.json";

    @Inject
    private ImposterConfig imposterConfig;

    @Inject
    private OpenApiService openApiService;

    private List<OpenApiPluginConfig> configs;

    /**
     * Holds the specifications.
     */
    private Cache<String, String> specCache = CacheBuilder.newBuilder().build();

    @Override
    protected Class<OpenApiPluginConfig> getConfigClass() {
        return OpenApiPluginConfig.class;
    }

    @Override
    protected void configurePlugin(List<OpenApiPluginConfig> configs) {
        this.configs = configs;
    }

    @Override
    public void configureRoutes(Router router) {
        final List<Swagger> allSpecs = Lists.newArrayListWithExpectedSize(configs.size());

        // specification mock endpoints
        configs.forEach(config -> {
            final Swagger swagger = new SwaggerParser().read(Paths.get(
                    config.getParentDir().getAbsolutePath(), config.getSpecFile()).toString());

            if (null != swagger) {
                allSpecs.add(swagger);
                swagger.getPaths().forEach((path, pathConfig) ->
                        handlePathOperations(router, config, swagger, path, pathConfig));

            } else {
                throw new RuntimeException(String.format("Unable to load API specification: %s", config.getSpecFile()));
            }
        });

        // serve specification and UI
        LOGGER.debug("Adding specification UI at: {}", SPECIFICATION_PATH);
        router.get(COMBINED_SPECIFICATION_PATH).handler(handleAsync(routingContext -> handleCombinedSpec(routingContext, allSpecs)));
        router.getWithRegex(SPECIFICATION_PATH + "$").handler(handleAsync(routingContext -> routingContext.response().putHeader("Location", SPECIFICATION_PATH + "/").setStatusCode(HttpUtil.HTTP_MOVED_PERM).end()));
        router.get(SPECIFICATION_PATH + "/*").handler(StaticHandler.create(UI_WEB_ROOT));
    }

    /**
     * Bind a handler to each operation.
     *
     * @param router     the Vert.x router
     * @param config     the plugin configuration
     * @param swagger    the OpenAPI specification
     * @param path       the mock path
     * @param pathConfig the path configuration
     */
    private void handlePathOperations(Router router, OpenApiPluginConfig config, Swagger swagger, String path, Path pathConfig) {
        pathConfig.getOperationMap().forEach((httpMethod, operation) -> {
            final String fullPath = ofNullable(swagger.getBasePath()).orElse("") + convertPath(path);
            LOGGER.debug("Adding mock endpoint: {} -> {}", httpMethod, fullPath);

            // convert an {@link io.swagger.models.HttpMethod} to an {@link io.vertx.core.http.HttpMethod}
            final HttpMethod method = HttpMethod.valueOf(httpMethod.name());
            router.route(method, fullPath).handler(buildHandler(config, swagger, operation));
        });
    }

    /**
     * Returns an OpenAPI specification combining all the given specifications.
     *
     * @param routingContext the Vert.x routing context
     * @param allSpecs       all specifications
     */
    private void handleCombinedSpec(RoutingContext routingContext, List<Swagger> allSpecs) {
        try {
            final String combinedJson = specCache.get("combinedSpec", () -> {
                try {
                    final Scheme scheme = Scheme.forValue(imposterConfig.getPluginArgs().get(ARG_SCHEME));
                    final String basePath = imposterConfig.getPluginArgs().get(ARG_BASEPATH);
                    final String title = imposterConfig.getPluginArgs().get(ARG_TITLE);

                    final Swagger combined = openApiService.combineSpecifications(allSpecs, basePath, scheme, title);
                    return MapUtil.MAPPER.writeValueAsString(combined);

                } catch (JsonGenerationException e) {
                    throw new ExecutionException(e);
                }
            });

            routingContext.response()
                    .putHeader(HttpUtil.CONTENT_TYPE, CONTENT_TYPE_JSON)
                    .end(combinedJson);

        } catch (Exception e) {
            routingContext.fail(e);
        }
    }

    /**
     * Convert the OpenAPI path to a Vert.x path, including any parameter placeholders.
     *
     * @param path the OpenAPI path
     * @return the Vert.x path
     */
    private String convertPath(String path) {
        boolean matchFound;
        do {
            final Matcher matcher = PATH_PARAM_PLACEHOLDER.matcher(path);
            matchFound = matcher.find();
            if (matchFound) {
                path = matcher.replaceFirst(":" + matcher.group(1));
            }
        } while (matchFound);

        return path;
    }

    /**
     * Build a handler for the given operation.
     *
     * @param config    the plugin configuration
     * @param swagger   the OpenAPI specification
     * @param operation the specification operation  @return a route handler
     */
    private Handler<RoutingContext> buildHandler(OpenApiPluginConfig config, Swagger swagger, Operation operation) {
        return handleAsync(routingContext -> {
            final HashMap<String, Object> context = Maps.newHashMap();
            context.put("operation", operation);

            scriptHandler(config, routingContext, getInjector(), context, responseBehaviour -> {
                final String statusCode = String.valueOf(responseBehaviour.getStatusCode());

                // look for a specification response based on the status code
                final Optional<Response> optionalMockResponse = operation.getResponses().entrySet().parallelStream()
                        .filter(mockResponse -> mockResponse.getKey().equals(statusCode))
                        .map(Map.Entry::getValue)
                        .findAny();

                // set status code regardless of response strategy
                final HttpServerResponse response = routingContext.response()
                        .setStatusCode(responseBehaviour.getStatusCode());

                if (optionalMockResponse.isPresent()) {
                    serveMockResponse(config, swagger, operation, routingContext, responseBehaviour, optionalMockResponse.get());

                } else {
                    LOGGER.info("No explicit mock response found for URI {} and status code {}",
                            routingContext.request().absoluteURI(), statusCode);

                    response.end();
                }
            });
        });
    }

    /**
     * Build a response from the specification.
     *
     * @param config            the plugin configuration
     * @param swagger           the OpenAPI specification
     * @param operation         the specification operation
     * @param routingContext    the Vert.x routing context
     * @param responseBehaviour the response behaviour
     * @param mockResponse      the specification response
     */
    private void serveMockResponse(OpenApiPluginConfig config, Swagger swagger, Operation operation,
                                   RoutingContext routingContext, ResponseBehaviour responseBehaviour,
                                   Response mockResponse) {

        LOGGER.trace("Found mock response for URI {} and status code {}",
                routingContext.request().absoluteURI(), responseBehaviour.getStatusCode());

        if (!responseBehaviour.getResponseHeaders().isEmpty()) {
            responseBehaviour.getResponseHeaders().forEach((header, value) ->
                    routingContext.response().putHeader(header, value));
        }

        if (!Strings.isNullOrEmpty(responseBehaviour.getResponseFile())) {
            // response file takes precedence
            serveResponseFile(config, routingContext, responseBehaviour);

        } else if (!Strings.isNullOrEmpty(responseBehaviour.getResponseData())) {
            // response data
            LOGGER.info("Response data is: {}", responseBehaviour.getResponseData());
            routingContext.response().putHeader(CONTENT_TYPE, CONTENT_TYPE_JSON)
                    .setStatusCode(responseBehaviour.getStatusCode()).end(responseBehaviour.getResponseData());

        } else {
            // attempt to serve an example from the specification, falling back if not present
            serveExample(config, swagger, routingContext, responseBehaviour, operation, mockResponse, this::fallback);
        }
    }

    /**
     * Reply with a static response file.
     *
     * @param config            the plugin configuration
     * @param routingContext    the Vert.x routing context
     * @param responseBehaviour the response behaviour
     */
    private void serveResponseFile(OpenApiPluginConfig config, RoutingContext routingContext,
                                   ResponseBehaviour responseBehaviour) {

        LOGGER.debug("Serving response file {} for URI {} and status code {}",
                responseBehaviour.getResponseFile(),
                routingContext.request().absoluteURI(),
                responseBehaviour.getStatusCode());

        final HttpServerResponse response = routingContext.response();

        // explicit content type
        if (!Strings.isNullOrEmpty(config.getContentType())) {
            response.putHeader(CONTENT_TYPE, config.getContentType());
        }

        response.sendFile(Paths.get(config.getParentDir().getAbsolutePath(),
                responseBehaviour.getResponseFile()).toString());
    }

    /**
     * Attempt to respond with an example from the API specification.
     *
     * @param config            the plugin configuration
     * @param swagger           the OpenAPI specification
     * @param routingContext    the Vert.x routing context
     * @param responseBehaviour the response behaviour
     * @param operation         the specification operation
     * @param mockResponse      the specification response
     * @param fallback          callback to invoke if no example was served
     */
    private void serveExample(OpenApiPluginConfig config, Swagger swagger, RoutingContext routingContext,
                              ResponseBehaviour responseBehaviour, Operation operation,
                              Response mockResponse, BiConsumer<RoutingContext, ResponseBehaviour> fallback) {

        final int statusCode = responseBehaviour.getStatusCode();

        @SuppressWarnings("unchecked") final Map<String, Object> examples = ofNullable(mockResponse.getExamples()).orElse(Collections.EMPTY_MAP);

        if (examples.size() > 0) {
            LOGGER.trace("Checking for mock example in specification ({} candidates) for URI {} and status code {}",
                    examples.size(), routingContext.request().absoluteURI(), statusCode);

            final Optional<Map.Entry<String, Object>> example = findExample(
                    routingContext, config, swagger, operation, examples);

            // serve example
            if (example.isPresent()) {
                final Map.Entry<String, Object> exampleEntry = example.get();

                final String exampleResponse = exampleEntry.getValue().toString();
                if (LOGGER.isTraceEnabled()) {
                    LOGGER.trace("Serving mock example for URI {} and status code {}: {}",
                            routingContext.request().absoluteURI(), statusCode, exampleResponse);
                } else {
                    LOGGER.info("Serving mock example for URI {} and status code {} (response body {} bytes)",
                            routingContext.request().absoluteURI(), statusCode,
                            ofNullable(exampleResponse).map(String::length).orElse(0));
                }

                // example key is its content type (should match one in the response 'provides' list)
                routingContext.response()
                        .putHeader(CONTENT_TYPE, exampleEntry.getKey())
                        .end(exampleResponse);

                return;
            }
        }

        LOGGER.trace("No example matches found in specification for URI {} and status code {}",
                routingContext.request().absoluteURI(), statusCode);

        // no matching example - use fallback
        fallback.accept(routingContext, responseBehaviour);
    }

    /**
     * Locate an example, first by searching the matched content types, then, optionally, the first found.
     *
     * @param routingContext the Vert.x routing context
     * @param config         the plugin configuration
     * @param swagger        the OpenAPI specification
     * @param operation      the specification operation    @return an optional example entry
     * @param examples       the specification response examples
     */
    private Optional<Map.Entry<String, Object>> findExample(RoutingContext routingContext, OpenApiPluginConfig config,
                                                            Swagger swagger, Operation operation,
                                                            Map<String, Object> examples) {

        // consolidate the produced content types
        final Set<String> produces = Sets.newHashSet();
        ofNullable(swagger.getProduces()).ifPresent(produces::addAll);
        ofNullable(operation.getProduces()).ifPresent(produces::addAll);

        // match accepted content types to those produced by this response operation
        final List<String> matchedContentTypes = HttpUtil.readAcceptedContentTypes(routingContext).parallelStream()
                .filter(produces::contains)
                .collect(Collectors.toList());

        // match first example by produced and accepted content types
        if (matchedContentTypes.size() > 0) {
            final Optional<Map.Entry<String, Object>> firstMatchingExample = examples.entrySet().parallelStream()
                    .filter(example -> matchedContentTypes.contains(example.getKey()))
                    .findFirst();

            if (firstMatchingExample.isPresent()) {
                final Map.Entry<String, Object> example = firstMatchingExample.get();
                LOGGER.debug("Exact example match found ({}) from specification", example.getKey());

                return Optional.of(example);
            }
        }

        // fallback to first example found
        if (config.isPickFirstIfNoneMatch()) {
            final Map.Entry<String, Object> example = examples.entrySet().iterator().next();
            LOGGER.debug("No exact example match found - choosing one example ({}) from specification." +
                    " You can switch off this behaviour by setting configuration option: pickFirstIfNoneMatch=false", example.getKey());

            return Optional.of(example);
        }

        // no matching example
        return empty();
    }

    /**
     * Handles the scenario when no example is found.
     *
     * @param routingContext    the Vert.x routing context
     * @param responseBehaviour the response behaviour
     */
    private void fallback(RoutingContext routingContext, ResponseBehaviour responseBehaviour) {
        LOGGER.warn("No example match found and no response file set for mock response for URI {} and status code {}" +
                " - sending empty response", routingContext.request().absoluteURI(), responseBehaviour.getStatusCode());

        routingContext.response().end();
    }
}

File: mock/openapi/src/main/java/io/gatehill/imposter/plugin/openapi/OpenApiPluginImpl.kt
/*
 * Copyright (c) 2016-2021.
 *
 * This file is part of Imposter.
 *
 * "Commons Clause" License Condition v1.0
 *
 * The Software is provided to you by the Licensor under the License, as
 * defined below, subject to the following condition.
 *
 * Without limiting other conditions in the License, the grant of rights
 * under the License will not include, and the License does not grant to
 * you, the right to Sell the Software.
 *
 * For purposes of the foregoing, "Sell" means practicing any or all of
 * the rights granted to you under the License to provide to third parties,
 * for a fee or other consideration (including without limitation fees for
 * hosting or consulting/support services related to the Software), a
 * product or service whose value derives, entirely or substantially, from
 * the functionality of the Software. Any license notice or attribution
 * required by the License must also include this Commons Clause License
 * Condition notice.
 *
 * Software: Imposter
 *
 * License: GNU Lesser General Public License version 3
 *
 * Licensor: Peter Cornish
 *
 * Imposter is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Imposter is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Imposter.  If not, see <https://www.gnu.org/licenses/>.
 */
package io.gatehill.imposter.plugin.openapi

import io.gatehill.imposter.ImposterConfig
import io.gatehill.imposter.http.HttpExchange
import io.gatehill.imposter.http.HttpExchangeHandler
import io.gatehill.imposter.http.HttpRouter
import io.gatehill.imposter.http.SingletonResourceMatcher
import io.gatehill.imposter.http.StatusCodeFactory
import io.gatehill.imposter.plugin.PluginInfo
import io.gatehill.imposter.plugin.RequireModules
import io.gatehill.imposter.plugin.config.ConfiguredPlugin
import io.gatehill.imposter.plugin.config.resource.BasicResourceConfig
import io.gatehill.imposter.plugin.config.resource.ResourceMethod
import io.gatehill.imposter.plugin.openapi.config.OpenApiPluginConfig
import io.gatehill.imposter.plugin.openapi.http.OpenApiResponseBehaviourFactory
import io.gatehill.imposter.plugin.openapi.model.ParsedSpec
import io.gatehill.imposter.plugin.openapi.service.ExampleService
import io.gatehill.imposter.plugin.openapi.service.SpecificationLoaderService
import io.gatehill.imposter.plugin.openapi.service.SpecificationService
import io.gatehill.imposter.script.ResponseBehaviour
import io.gatehill.imposter.server.ServerFactory
import io.gatehill.imposter.service.ResourceService
import io.gatehill.imposter.service.ResponseRoutingService
import io.gatehill.imposter.service.ResponseService
import io.gatehill.imposter.service.ResponseService.ResponseSender
import io.gatehill.imposter.util.HttpUtil
import io.gatehill.imposter.util.LogUtil.describeRequestShort
import io.gatehill.imposter.util.MapUtil.addJavaTimeSupport
import io.gatehill.imposter.util.ResourceUtil
import io.gatehill.imposter.util.ResourceUtil.convertPathFromOpenApi
import io.swagger.util.Json
import io.swagger.v3.oas.models.OpenAPI
import io.swagger.v3.oas.models.Operation
import io.swagger.v3.oas.models.PathItem
import io.swagger.v3.oas.models.media.Content
import io.swagger.v3.oas.models.responses.ApiResponse
import io.vertx.core.Vertx
import org.apache.logging.log4j.LogManager
import javax.inject.Inject
import kotlin.collections.component1
import kotlin.collections.component2
import kotlin.collections.set

/**
 * Plugin for OpenAPI (OAI; formerly known as 'Swagger').
 *
 * @author Pete Cornish
 */
@PluginInfo("openapi")
@RequireModules(OpenApiModule::class)
class OpenApiPluginImpl @Inject constructor(
    vertx: Vertx,
    imposterConfig: ImposterConfig,
    private val resourceService: ResourceService,
    private val specificationService: SpecificationService,
    private val exampleService: ExampleService,
    private val responseService: ResponseService,
    private val responseRoutingService: ResponseRoutingService,
    private val openApiResponseBehaviourFactory: OpenApiResponseBehaviourFactory,
    private val serverFactory: ServerFactory,
    private val specificationLoaderService: SpecificationLoaderService,
) : ConfiguredPlugin<OpenApiPluginConfig>(
    vertx, imposterConfig
) {
    override val configClass = OpenApiPluginConfig::class.java
    private lateinit var allSpecs: List<ParsedSpec>

    companion object {
        private val LOGGER = LogManager.getLogger(OpenApiPluginImpl::class.java)
        private const val UI_WEB_ROOT = "swagger-ui"
        private const val ARG_BASEPATH = "openapi.basepath"

        /**
         * 'default' is a special case in OpenAPI that does not have a status code.
         */
        private const val DEFAULT_RESPONSE_KEY = "default"
        const val SPECIFICATION_PATH = "/_spec"
        const val COMBINED_SPECIFICATION_PATH = "$SPECIFICATION_PATH/combined.json"

        init {
            addJavaTimeSupport(Json.mapper())
        }
    }

    private val basePath: String? by lazy {
        imposterConfig.pluginArgs!![ARG_BASEPATH]
    }

    private val resourceMatcher = SingletonResourceMatcher.instance

    override fun configureRoutes(router: HttpRouter) {
        if (configs.isEmpty()) {
            LOGGER.debug("No OpenAPI configuration files provided - skipping plugin setup")
            return
        }

        parseSpecs(router)

        // serve specification and UI
        LOGGER.debug("Adding specification UI at: {}{}", imposterConfig.serverUrl, SPECIFICATION_PATH)
        router.get(COMBINED_SPECIFICATION_PATH).handler(
            resourceService.handleRoute(imposterConfig, configs, resourceMatcher) { httpExchange: HttpExchange ->
                handleCombinedSpec(httpExchange)
            }
        )
        router.getWithRegex("$SPECIFICATION_PATH$").handler(
            resourceService.handleRoute(imposterConfig, configs, resourceMatcher) { httpExchange: HttpExchange ->
                httpExchange.response()
                    .putHeader("Location", "$SPECIFICATION_PATH/")
                    .setStatusCode(HttpUtil.HTTP_MOVED_PERM)
                    .end()
            }
        )
        router.get("$SPECIFICATION_PATH/*").handler(serverFactory.createStaticHttpHandler(UI_WEB_ROOT))
    }

    private fun parseSpecs(router: HttpRouter) {
        val parsedSpecs = mutableListOf<ParsedSpec>()

        // specification mock endpoints
        configs.forEach { config: OpenApiPluginConfig ->
            val spec = specificationLoaderService.parseSpecification(config)

            // The *path prefix* includes the plugin configuration root path,
            // but not the server 'basePath', as this is required only for
            // the serving prefix and the list of server entries added to
            // the combined spec. Crucially, the path prefix is incorporated
            // into each of the paths within the spec.
            val pathPrefix = (config.path ?: "")

            // The *serving prefix* includes the spec's first server entry path, however,
            // the path in the combined spec document should not include this, as server
            // entries are automatically prefixed by the spec UI.
            //
            // It is built from a concatenation of:
            // 1. the server 'basePath'
            // 2. the plugin configuration root path
            // 3. the path of the first 'server' entry in the spec
            val servingPrefix = (basePath ?: "") + pathPrefix + if (config.stripServerPath) "" else specificationService.determinePathFromSpec(spec)

            spec.paths.forEach { path: String, pathConfig: PathItem ->
                handlePathOperations(router, config, spec, servingPrefix, path, pathConfig)
            }
            parsedSpecs += ParsedSpec(spec, pathPrefix)
        }

        allSpecs = parsedSpecs
    }

    /**
     * Bind a handler to each operation.
     *
     * @param router     the Vert.x router
     * @param config     the plugin configuration
     * @param spec       the OpenAPI specification
     * @param pathPrefix
     * @param path       the mock path
     * @param pathConfig the path configuration
     */
    private fun handlePathOperations(
        router: HttpRouter,
        config: OpenApiPluginConfig,
        spec: OpenAPI,
        pathPrefix: String,
        path: String,
        pathConfig: PathItem
    ) {
        pathConfig.readOperationsMap().forEach { (httpMethod: PathItem.HttpMethod, operation: Operation) ->
            val fullPath = buildFullPath(pathPrefix, path)
            LOGGER.debug("Adding mock endpoint: {} -> {}", httpMethod, fullPath)

            // convert an io.swagger.models.HttpMethod to an io.vertx.core.http.HttpMethod
            val method = ResourceMethod.valueOf(httpMethod.name)
            router.route(method, fullPath).handler(buildHandler(config, operation, spec))
        }
    }

    /**
     * Construct the full path from the base path and the operation path.
     *
     * @param pathPrefix
     * @param specOperationPath the operation path from the OpenAPI specification
     * @return the full path
     */
    private fun buildFullPath(pathPrefix: String, specOperationPath: String): String {
        val operationPath = convertPathFromOpenApi(specOperationPath)

        return if (pathPrefix.endsWith("/")) {
            if (operationPath!!.startsWith("/")) {
                pathPrefix + operationPath.substring(1)
            } else {
                pathPrefix + operationPath
            }
        } else {
            if (operationPath!!.startsWith("/")) {
                pathPrefix + operationPath
            } else {
                "$pathPrefix/$operationPath"
            }
        }
    }

    /**
     * Returns an OpenAPI specification combining all the given specifications.
     *
     * @param httpExchange the HTTP exchange
     */
    private fun handleCombinedSpec(httpExchange: HttpExchange) {
        try {
            httpExchange.response()
                .putHeader(HttpUtil.CONTENT_TYPE, HttpUtil.CONTENT_TYPE_JSON)
                .end(specificationService.getCombinedSpecSerialised(allSpecs, basePath))
        } catch (e: Exception) {
            httpExchange.fail(e)
        }
    }

    /**
     * Build a handler for the given operation.
     *
     * @param pluginConfig the plugin configuration
     * @param operation    the specification operation
     * @param spec         the OpenAPI specification
     * @return a route handler
     */
    private fun buildHandler(
        pluginConfig: OpenApiPluginConfig,
        operation: Operation,
        spec: OpenAPI
    ): HttpExchangeHandler {
        // statically calculate as much as possible
        val statusCodeFactory = buildStatusCodeCalculator(operation)
        return resourceService.handleRoute(imposterConfig, pluginConfig, resourceMatcher) { httpExchange: HttpExchange ->
            LOGGER.trace("Operation ${operation.operationId} matched for request: ${describeRequestShort(httpExchange)}")

            if (!specificationService.isValidRequest(pluginConfig, httpExchange, allSpecs, basePath)) {
                return@handleRoute
            }

            val context = mutableMapOf<String, Any>()
            context["operation"] = operation

            val request = httpExchange.request()
            val resourceConfig = httpExchange.get<BasicResourceConfig>(ResourceUtil.RESOURCE_CONFIG_KEY)

            val defaultBehaviourHandler = { responseBehaviour: ResponseBehaviour ->
                // set status code regardless of response strategy
                val response = httpExchange.response().setStatusCode(responseBehaviour.statusCode)

                findApiResponse(operation, responseBehaviour.statusCode)?.let { specResponse ->
                    if (!responseBehaviour.responseHeaders.containsKey(HttpUtil.CONTENT_TYPE)) {
                        setContentTypeFromSpec(httpExchange, responseBehaviour, specResponse)
                    }

                    // build a response from the specification
                    val exampleSender =
                        ResponseSender { httpExchange: HttpExchange, responseBehaviour: ResponseBehaviour ->
                            exampleService.serveExample(
                                imposterConfig,
                                pluginConfig,
                                httpExchange,
                                responseBehaviour,
                                specResponse,
                                spec
                            )
                        }

                    // attempt to serve an example from the specification, falling back if not present
                    responseService.sendResponse(
                        pluginConfig,
                        resourceConfig,
                        httpExchange,
                        responseBehaviour,
                        exampleSender,
                        ResponseSender { httpExchange: HttpExchange, responseBehaviour: ResponseBehaviour ->
                            fallback(httpExchange, responseBehaviour)
                        })
                } ?: run {
                    LOGGER.warn(
                        "No response found in specification for {} {} and status code {}",
                        request.method(),
                        request.path(),
                        responseBehaviour.statusCode
                    )
                    response.end()
                }
            }

            responseRoutingService.route(
                pluginConfig,
                resourceConfig,
                httpExchange,
                context,
                statusCodeFactory,
                openApiResponseBehaviourFactory,
                defaultBehaviourHandler
            )
        }
    }

    private fun setContentTypeFromSpec(
        httpExchange: HttpExchange,
        responseBehaviour: ResponseBehaviour,
        optionalResponse: ApiResponse
    ) {
        optionalResponse.content?.let { responseContent: Content ->
            responseContent.keys.firstOrNull()?.let { firstContentType ->
                when (responseContent.size) {
                    0 -> return
                    1 -> LOGGER.debug(
                        "Setting content type [{}] from specification for {}",
                        firstContentType,
                        describeRequestShort(httpExchange)
                    )
                    else -> LOGGER.warn(
                        "Multiple content types in specification - selecting first [{}] for {}",
                        firstContentType,
                        describeRequestShort(httpExchange)
                    )
                }
                responseBehaviour.responseHeaders[HttpUtil.CONTENT_TYPE] = firstContentType
            }
        }
    }

    private fun buildStatusCodeCalculator(operation: Operation) = StatusCodeFactory { rc: BasicResourceConfig ->
        rc.responseConfig.statusCode
        // Fall back to the first non-default response for this operation.
        // Note: responses are keyed on their status code and "default" is a
        // special case in OpenAPI that does not have a status code.
            ?: operation.responses?.keys?.firstOrNull { !DEFAULT_RESPONSE_KEY.equals(it, ignoreCase = true) }?.toInt()
            ?: HttpUtil.HTTP_OK
    }

    /**
     * @return an API response for the given [statusCode], or `null`
     */
    private fun findApiResponse(operation: Operation, statusCode: Int): ApiResponse? {
        // openapi statuses are represented as strings
        val status = statusCode.toString()

        // look for a specification response based on the status code
        val optionalResponse = operation.responses?.filter { it.key == status }?.values?.firstOrNull()

        return optionalResponse ?: run {
            // fall back to default (might also be null)
            LOGGER.debug(
                "No response found for status code {}; falling back to default response if present",
                statusCode
            )
            operation.responses.default
        }
    }

    /**
     * Handles the scenario when no example is found.
     *
     * @param httpExchange    the HTTP exchange
     * @param responseBehaviour the response behaviour
     */
    private fun fallback(httpExchange: HttpExchange, responseBehaviour: ResponseBehaviour): Boolean {
        LOGGER.warn(
            "No example match found and no response file set for mock response for URI {} with status code {}" +
                    " - sending empty response", httpExchange.request().absoluteURI(), responseBehaviour.statusCode
        )
        httpExchange.response().end()
        return true
    }
}